apollo-gradle-plugin-incubating/src/main/kotlin/com/apollographql/apollo/gradle/ServiceVariant.kt

```diff
@@ -0,0 +1,143 @@
+package com.apollographql.apollo.gradle
+
+import com.apollographql.apollo.compiler.*
+import org.gradle.api.Project
+import org.gradle.api.tasks.util.PatternSet
+import java.io.File
+
+class ServiceVariant(
+    val name: String,
+    val files: List<File>,
+    val schemaFile: File?,
+    val schemaPackageName: String,
+    val rootPackageName: String
+) {
+
+  companion object {
+    fun from(project: Project, sourceSetNames: List<String>, service: Service): ServiceVariant {
+      val schemaFilePath = service.schemaFilePath
+      if (schemaFilePath == null) {
+        throw IllegalArgumentException("Please define schemaFilePath for service '${service.name}'")
+      }
+
+      var schemaFile = project.projectDir.child(schemaFilePath)
+      if (!schemaFile.isFile) {
+        throw IllegalArgumentException("No schema found at:\n${schemaFile.absolutePath}")
+      }
+      val schemaKey = schemaFile.canonicalPath.relativePathToGraphql()
+      if (schemaKey != null) {
+        // schema is under src/{foo}/graphql, see if it is overriden by anoter variant
+        val schemaCandidates = findFilesInSourceSets(project, sourceSetNames, schemaKey) { true }
+        schemaFile = schemaCandidates.values.first()
+      }
+
+      val sourceFolderPath = if (service.sourceFolderPath != null) {
+        service.sourceFolderPath!!
+      } else {
+        // if schemaFilePath is outside src/{foo}/graphql, search the whole graphql folder
+        schemaFile.canonicalPath.relativePathToGraphql(dropLast = 1) ?: "."
+      }
+
+      val candidateFiles = findFilesInSourceSets(project, sourceSetNames, sourceFolderPath, ::isGraphQL).values.toList()
+
+      val files = if (service.exclude != null) {
+        val patternSet = PatternSet()
+        patternSet.exclude(service.exclude!!)
+        project.files(candidateFiles).asFileTree.matching(patternSet).toList()
+      } else {
+        candidateFiles
+      }
+
+      val schemaPackageName = schemaFile.canonicalPath.formatPackageName(dropLast = 1) ?: ""
+
+      return ServiceVariant(
+          name = service.name,
+          files = files,
+          schemaFile = schemaFile,
+          schemaPackageName = schemaPackageName,
+          rootPackageName = service.rootPackageName ?: ""
+      )
+    }
+
+    fun default(project: Project, sourceSetNames: List<String>): List<ServiceVariant> {
+      val schemaFiles = findFilesInSourceSets(project, sourceSetNames, ".") {
+        it.name == "schema.json"
+      }
+
+      if (schemaFiles.isEmpty()) {
+        return emptyList()
+      }
+
+      var i = 0
```

Comment:
I see it used only in one place to format name of service variant. Should we just use hardcoded `Default` as a name instead?

Reply:
It can't be just `Default` because there can be several services. For an example:

```
com/githunt/schema.json
com/githunt/queries.graphql
com/starwars/schema.json
com/starwars/queries.graphql
```

The other option (that is commented out at the moment) is to use the packageName as identifier for the service (i.e `generateDebugComGithunApolloClasses` instead of `generateDebug0ApolloClasses`).